### DISCUSSION TITLE
How to interop with NSWindow in Avalonia?

### DISCUSSION THREAD

**Discussion by BreyerW** [AvaloniaUI/Avalonia#13174] [Q&A]
Im completely new to Apple interop can someone tell me how to define `DllImport` for `NSWindow.orderedIndex`? i have native `IntPtr` to window handed out by Avalonia but now i need to get its zIndex. I managed to interop with Window and X11 equivalents but i have 0 idea how to start with obj c and NSWindow. Alternatively someone has guide/resources for obj c interop? Tried looking myself but resources specifically for obj c -> c# interop are scarce.

**Comment by maxkatz6** [comment]
There are several approaches:

1. Easiest way: use "net7.0-macos" target and then you can access [NSWindow](https://learn.microsoft.com/en-us/dotnet/api/appkit.nswindow?view=xamarin-mac-sdk-14) classes. Probably the most advisable way for applications.

2. x64 way (and rosetta, no m1 support) - use [MonoMac](https://www.nuget.org/packages/MonoMac.NetStandard) dependency. Easiest, but limited.

3. Native dependency - build Objective C dylib and reference it from C# code. Avalonia uses this approach with our macOS backend. I would say it's the best approach, if you are experienced with C++/C, and are able to ship native binaries with the app.

4. The hardest approach, but one without any dependencies (native or managed) - pinvoke. Which is very complicated in the context of apple interop. Let's get `orderedIndex` as an example:
https://developer.apple.com/documentation/appkit/nswindow/1449577-orderedindex?language=objc
In obj c this property would be set in this way:
```objc
[window setOrderedIndex: 0];
```
You already have a pointer to `window` object from Avalonia. 
Then you need to follow these steps:
1. Create a selector for `setOrderedIndex` method (using [NSSelectorFromString](https://developer.apple.com/documentation/foundation/1395294-nsselectorfromstring?language=objc) pinvoke). Cache it.
2. Create `objc_msgSend ` PInvoke with three parameters: (IntPtr windowPtr, IntPtr setOrderedIndexSelector, Int orderIndex) returning void.
3. Call this sobjc_msgSend. 

This StackOverflow answer is very helpful as well:
https://stackoverflow.com/questions/44665544/how-to-pinvoke-appkit-methods-from-net-core-application
Also:
https://github.com/dotnet/runtime/blob/26a809adb958c604ac773d39f2c253cf0dd12e1e/src/libraries/Common/src/Interop/OSX/Interop.libobjc.cs
https://www.mono-project.com/archived/monomac_documentation_api_design/ (more about theory)

If your library/application only needs this one setOrderedIndex call, might be worth it to try the last approach with PInvokes.

**Reply by BreyerW** [reply]
Ohh thanks for detailed reply. That being said i actually need getter not setter (im developing Multiple Document Interface like Avalon.Dock but with ability to drag and drop into windows that arent origin of dragged tab so i need zindex to correctly detect which windows are on top under mouse - afaik no other library in avalonia ecosystem offers this).

Im assuming correctly i can change to getOrderedIndex and change parameter to be return of getter? And yes i need only this property nothing else.

**Reply by Sathiyaraman-M** [reply]
@maxkatz6 Can you please elaborate more on the first one, the easiest approach? Becuz, I had hard time doing that in my project, getting `AppKitThreadAccessException` while trying to access the `MainWindow` object, which is actually of `NSWindow` type.

**Reply by maxkatz6** [reply]
@BreyerW for `get` selectors they don't add `get`, so it just will be `orderedIndex`. I don't remember exactly, but you also need to add ":" in the end of selectors with parameters (so, `setOrderedIndex:`, but `orderedIndex` as there is no parameters).

@Sathiyaraman-M 

> while trying to access the MainWindow

I don't know if this is set in Avalonia, but you can try. Normally you can get NSWindow from Avalonia window itself (window.TryGetPlatformHandle().Handle, and then create `NSWindow` object from that handle). That's if you are using "net7.0-macos" target approach.

But I don't know why you are getting `AppKitThreadAccessException`. If your app is initialized on the main (very first) thread of the process, and you try to get MainWindow on the same thread, then in theory it should work just fine. With both AppKit and Avalonia dispatchers.

**Reply by maxkatz6** [reply]
@BreyerW official documentation (both MS and Apple) is lacking, so I often google methods/selectors usage in Objective C language, so it's easier to see how it should be ported to .NET. Swift examples often add too much syntax sugar.
![image](https://github.com/AvaloniaUI/Avalonia/assets/3163374/5e2ecc4b-a8ef-4c2f-b858-1c443e7ee8f1)

Also, with more advanced scenarios, it gets much more complicated when you need to pass [block methods](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/WorkingwithBlocks/WorkingwithBlocks.html) (aka lambda methods, callbacks). Since you can't just pass a function pointer, I would recommend using other approaches instead. 
But with simple scenarios like orderedIndex/setOrderedIndex you don't have that problem.
